src/full/Agda/TypeChecking/Sort.hs

```diff
@@ -124,18 +136,35 @@ ptsRule' a b c = do
     (equalSort c' c)
 
 hasPTSRule :: Dom Type -> Abs Sort -> TCM ()
-hasPTSRule a b = void $ inferPiSort a b
+hasPTSRule a b =
+  if alwaysValidCodomain $ unAbs b
+  then return ()
+  else do
+    infer <- reduce =<< instantiateFull =<< inferPiSort a b
+    case infer of
+      PiSort a s1 s2  | noMetas s1 && noMetas (unAbs s2) -> typeError $ InvalidTypeSort infer
+      FunSort s1 s2   | noMetas s1 && noMetas s2 -> typeError $ InvalidTypeSort infer
```

Comment:
I think I see what you are doing here: if we get a `PiSort` or a `FunSort` with no metas, then that means it will never reduce to anything, so there is no PTS rule. I think a better way to produce the same effect would be to use `reduceB` and check if the result is `NotBlocked` or `Blocked` with blocking tag `neverUnblock`: both of these also mean that the sort will never reduce further, but with the advantage that you can get rid of the call to `instantiateFull` above.